kitsune/wiki/utils.py

```diff
@@ -194,3 +212,51 @@ def get_visible_document_or_404(
 
 def get_visible_revision_or_404(user, **kwargs):
     return get_object_or_404(Revision.objects.visible(user, **kwargs))
+
+
+def build_topics_data(request, product, topics):
+    """Build topics_data for use in topic cards"""
+    topics_data = []
+
+    # Get all documents for all topics up front
+    all_docs, _ = documents_for(
+        request.user, request.LANGUAGE_CODE, topics=list(topics), products=[product]
+    )
+
+    # Convert all docs to Document objects
+    doc_ids = [doc["id"] for doc in all_docs]
+    documents = (
+        Document.objects.filter(id__in=doc_ids)
+        .prefetch_related("topics")
+        .annotate(
+            topic_ids=StringAgg(
+                Cast("topics__id", TextField()), delimiter=",", ordering="topics__id", default=""
+            )
+        )
+    )
+
+    for topic in topics:
+        # Get featured articles first (returns up to 4 Document objects)
+        featured_articles = get_featured_articles(
+            product, locale=request.LANGUAGE_CODE, topic=topic
+        )
+
+        # Get docs for this topic from documents
+        topic_docs = [
+            doc for doc in documents if doc.topic_ids and str(topic.id) in doc.topic_ids.split(",")
+        ]
+
+        # Create a generator for remaining docs
```

Comment:
Similar comment with CSS, the comment is not needed here. This applies to all submitted code.

Reply:
Trimmed out some of the more obvious items.